lib/src/client/proxy.dart

```diff
@@ -0,0 +1,23 @@
+class Proxy {
```

Comment:
I would just use `Proxy?` and `null` to denote direct connection instead of doing `Proxy.direct` and with a bunch of `null` / magical value fields. To avoid `!` everywhere you could do (for example):

```dart
Future<Stream<int>> connectImpl() async {
  socket = await Socket.connect(connectionHost, connectionPort, timeout: _options.connectTimeout);
  final proxy = _proxy;
  if (proxy == null) {
    return socket;
  }
  
  // We need to ask the proxy to connect to our destination host.
  return connectThroughProxy(proxy); // Have connectThroughProxy take an instance of proxy.
}
```